p2p/src/net/mod.rs

```diff
@@ -165,3 +182,11 @@ where
     /// Poll syncing-related event from the networking service
     async fn poll_next(&mut self) -> crate::Result<types::SyncingEvent<T>>;
 }
+
+// TODO: We can require infallible conversion after getting rid of `libp2p::MultiAddr`.
+/// Represents a fallible conversion to a bannable address.
+pub trait AsBannableAddress {
+    type BannableAddress;
+
+    fn as_bannable(&self) -> Option<Self::BannableAddress>;
```

Comment:
This shouldn't return an Option... what are we saying, there are addresses we can never ban? Either the whole design around banning is wrong or there's something major we don't understand.

Reply:
Technically `MultiAddr` can contain no IP address. For example, the following is a type-wise correct libp2p address: `/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N` as well as `Multiaddr::empty()`.

Reply:
But you see, that shouldn't be allowed. We have to think about this. ANY address coming to p2p should be bannable.

Reply:
In the current implementation such addresses are rejected during connection.

Reply:
Then we add a test for that, and consider this an invariant, and make the return a value that we panic if not satisfied.